GA update for module-design.mdx (#329)
hashicorp/terraform-docs-common: website/docs/cloud-docs/no-code-provisioning/module-design.mdx

---
page_title: Designing No-Code Ready Modules - No-Code Provisioning - Terraform Cloud
description: No-code ready modules let users deploy a module's resources without writing configuration. Prepare modules for no-code provisioning.
tfc_only: true
---

# Designing No-Code Ready Modules

> -> **Note:** No-code provisioning is available in the [Terraform Cloud Business tier](https://www.hashicorp.com/products/terraform/pricing).

Reviewer: (Yes it is, so I'm resolving this comment!)

- Deploying all resources needed for a three-tier web application
- Deploying a database with constraints on resource allocation and deployment region

### Adjust for New Module Versions

> No-code provisioning desgination is now pinned to a selected version of a module. You must select the version at the time of enabling the module. If no selection is made, it will default to the latest version at the time of enabling. This ensures that you can define variable input options for the intended version and that no-code users provision the most appropriate version. When a newer version of the module becomes avialable, you may want to update the pinned no-code version to the latest.

Reviewer 2: ```suggestion
No-code provisioning designation is pinned to a selected version of a module. You must select the version at the time of enabling the module. If no selection is made, it will default to the latest version at the time of enablement. This ensures that you can define variable input options for the intended version and that no-code users provision the most appropriate version. When a newer version of the module becomes available, you may decide to update the pinned no-code version to the latest.
```

Reviewer: @opslivia Did you see my suggestion here? If this doesn't feel good I can take another try at rewriting what is currently there! (I'm unresolving this comment for now)

Reviewer: (Closing because I resuggested these in a new review)

### Provide Variable Defaults When Possible

> The no-code provisioning workflow prompts users to set values for the module version's input variables before creating the new workspace and deploying resources. We recommend setting reasonable defaults when possible to reduce the effort and expertise needed to deploy the module. Remember that the workspace can also access variable values set through global variable sets in your organization.

Reviewer 2: ```suggestion
The no-code provisioning workflow prompts users to set values for the module version's input variables before creating the new workspace and deploying resources. We recommend setting reasonable defaults when possible to reduce the effort and expertise needed to deploy the module. Remember that the workspace can also access variable values set through global or project level variable sets in your organization.
```

### Define Dropdown Options for Variables without Defaults

> For variables without defaults, you may still want to limit what the no-code user can input. Define these input options at the time of enabling the module for no-code provisioning. You can do this either from the Terraform Cloud UI, via the [No-Code Provisioing API](), or via the [TFE provider](). The latter method is recommended for those that want to track these variable input options through a code approval process, keeping the configuration as code.

Reviewer 2: ```suggestion
For variables without defaults, you may still want to limit what the no-code user can input. Define these input options at the time of enabling the module for no-code provisioning. You can do this either from the Terraform Cloud UI, via the [No-Code Provisioing API](https://developer.hashicorp.com/terraform/cloud-docs/api-docs/no-code-provisioning), or via the [TFE provider](https://registry.terraform.io/providers/hashicorp/tfe/latest/docs/resources/no_code_module). The latter method is recommended for those that want to track these variable input options through a code approval process, keeping the configuration as code.
```

Reviewer: @opslivia (Un-resolving this thread too) Did you see my suggestion above?

Reviewer: (Closing because I resuggested these in a new review)

- Deploying all resources needed for a three-tier web application
- Deploying a database with constraints on resource allocation and deployment region

> ### Adjust for New Module Versions

Author: I used "Adjust for new" here because it's something they may do periodically over time, as new versions of a module get published. When a new version becomes available, the admins managing no-code status for a module will need to decide if they want to change the version pin and update the variable's input options.

Is there a better word to use here to describe this periodic maintenance?

> ### Adjust for New Module Versions

Reviewer: Thank you for the context here! I think the word "update" might be a better fit here if that is what admins will most commonly be doing? How does this feel?
```suggestion
### Updating a Module's version
```

> No-code provisioning designation is pinned to a selected version of a module. You must select the version at the time of enabling the module. If no selection is made, it will default to the latest version at the time of enablement. This ensures that you can define variable input options for the intended version and that no-code users provision the most appropriate version. When a newer version of the module becomes available, you may decide to update the pinned no-code version to the latest.

Reviewer: Re-suggesting:
```suggestion

When you enable no-code provisioning for a module, Terraform Cloud pins the **No-code Ready** designation to a specific module version of your choice. Terraform Cloud deploys the selected module version whenever a user provisions a workspace using that module.

By default, a module selects the latest version available. If you pinned a specific module version and a newer one becomes available, you can always update your module's version.

One of the advantages of using different module versions is that you can define input options for each variable in alternative versions. Another advantage is that your no-code users consistently provision workspaces with the correct module version. 
```

Author: I like the first two paragraphs. However, the last one doesn't make sense as worded. They won't be able to enable no-code for many versions at once.

Reviewer: ```suggestion

When you enable no-code provisioning for a module, Terraform Cloud pins the **No-code Ready** designation to the specific module version of your choice. Terraform Cloud deploys that selected module version whenever a user provisions a workspace using that module, ensuring that no-code users always provision the correct version. Pinning the **No-code Ready** designation to a specific module version lets you set variable input options, which are tied to that specific module version.

By default, a module selects the latest version available. If you pinned a specific module version and a newer one becomes available, you can always update your module's version.
```

### Provide Variable Defaults When Possible

The no-code provisioning workflow prompts users to set values for the module version's input variables before creating the new workspace and deploying resources. We recommend setting reasonable defaults when possible to reduce the effort and expertise needed to deploy the module. Remember that the workspace can also access variable values set through global or project level variable sets in your organization.

### Define Dropdown Options for Variables without Defaults

> For variables without defaults, you may still want to limit what the no-code user can input. Define these input options at the time of enabling the module for no-code provisioning. You can do this either from the Terraform Cloud UI, via the [No-Code Provisioing API](https://developer.hashicorp.com/terraform/cloud-docs/api-docs/no-code-provisioning), or via the [TFE provider](https://registry.terraform.io/providers/hashicorp/tfe/latest/docs/resources/no_code_module). The latter method is recommended for those that want to track these variable input options through a code approval process, keeping the configuration as code.

Reviewer: ```suggestion
If your module has variables without defaults, you can define options to limit the values a user can input while enabling your module for no-code provisioning. You can define input options using the Terraform Cloud UI, the [No-Code Provisioning API](/terraform/cloud-docs/api-docs/no-code-provisioning), or the [TFE provider](https://registry.terraform.io/providers/hashicorp/tfe/latest/docs/resources/no_code_module). We recommend using the TFE provider if you want to track variable input options through a code approval process, therefore keeping your configuration as code.
```

> Subsequent changes to the variable input options will be surfaced to new provisions of the module version. If you update the selected module version enabled for no-code provisioning, you may also want to revisit the variables and adjust the defined input values accordingly. 

Reviewer: I tweaked this to be in active voice. I think it is still correct😅, but please let me know if not:
```suggestion
Any changes to a variable's input options are surfaced when you provision a new module version. If you update the selected module version enabled for no-code provisioning, consider revisiting the variables and adjusting the defined input values accordingly. 
```

Author: As this doc is intended for the Admin or module author, I'd like to use part of your suggestion:

Any subsequent changes to a variable's input options are surfaced when no-code users provision a new module version. If you update the selected module version enabled for no-code provisioning, consider revisiting the variables and adjusting the defined input values accordingly.

Reviewer: Here is the suggestion you'd like added! This looks good to me too 👍 
```suggestion
Any subsequent changes to a variable’s input options are surfaced when no-code users provision a new module version. If you update the selected module version enabled for no-code provisioning, consider revisiting the variables and adjusting the defined input values accordingly.
```
